### PR TITLE
feat(clp-s): Add boilerplate for SQL parsing. 

### PR DESCRIPTION
**Pull request by gibber9809**
# Description
This PR adds boilerplate for the new sql parser. This includes a stubbed out g4 grammar and visitor, as well as a basic unit test that tells us that the whole thing compiles and runs.

# Validation performed
- validated that unit test linking against sql parser library builds and runs successfully

## Summary by CodeRabbit

- **New Features**
	- Introduced SQL parsing capabilities with a new SQL parser and grammar.
	- Added a `LogSuppressor` class to manage logging during tests.
	- Implemented unit tests for SQL parsing functionality, including handling edge cases.

- **Bug Fixes**
	- Enhanced error handling in SQL parsing to ensure proper logging and feedback.

### CODE REVIEW

File: components/core/src/clp_s/search/sql/sql.cpp [hunk 1]
```diff
@@ -0,0 +1,69 @@
+#include <iostream>
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/sql/sql.cpp:1]:
```suggestion
#include <any>
#include <iostream>
```

File: components/core/src/clp_s/search/sql/sql.cpp [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    CommonTokenStream tokens(&lexer);
+    SqlParser parser(&tokens);
+    parser.removeErrorListeners();
+    parser.addErrorListener(&parser_error_listener);
+    SqlParser::StartContext* tree = parser.start();
+
+    if (lexer_error_listener.error()) {
+        SPDLOG_ERROR("Lexer error: {}", lexer_error_listener.message());
+        return {};
+    } else if (parser_error_listener.error()) {
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/sql/sql.cpp:61]:
```suggestion
    }
    if (parser_error_listener.error()) {
```

File: components/core/src/clp_s/search/sql/sql.hpp [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+#define CLP_S_SEARCH_SQL_SQL_HPP
+
+#include <istream>
+
+#include "../Expression.hpp"
+
+namespace clp_s::search::sql {
+/**
+ * Generate a search AST from an SQL expression in an input stream
+ * @param in input stream containing an SQL expression followed by EOF
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/sql/sql.hpp:11]:
```suggestion
 * Parses an SQL expression from the given stream to generate a search AST.
 * @param in Input stream containing an SQL expression followed by EOF
```

File: components/core/tests/test-sql.cpp [hunk 4]
```diff
@@ -0,0 +1,27 @@
+#include <sstream>
+#include <string>
+#include <vector>
```
Comment (LinZhihao-723) [inline components/core/tests/test-sql.cpp:3]:
```suggestion
```

File: components/core/tests/test-sql.cpp [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+#include "../src/clp_s/search/EmptyExpr.hpp"
+#include "../src/clp_s/search/sql/sql.hpp"
+#include "LogSuppressor.hpp"
+
+using clp_s::search::EmptyExpr;
+using clp_s::search::sql::parse_sql_expression;
+using std::string;
+using std::stringstream;
+using std::vector;
```
Comment (LinZhihao-723) [inline components/core/tests/test-sql.cpp:16]:
```suggestion
using std::stringstream;
```

**Review by LinZhihao-723 — CHANGES_REQUESTED** [review]
Minor fixes

File: components/core/src/clp_s/search/sql/sql.cpp [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+private:
+    bool m_error{false};
+    std::string m_error_message;
+};
+
+class ParseTreeVisitor : public SqlBaseVisitor {
+public:
+    std::any visitStart(SqlParser::StartContext* ctx) override { return EmptyExpr::create(); }
+};
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/sql/sql.cpp:43]: Sorry to miss this in the last round of review. Shall we move these two classes into the anon namespace? Or if there's plan to reuse them in other places, we should probably move them into the dedicated headers.

File: components/core/src/clp_s/search/sql/sql.cpp [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+#include "../EmptyExpr.hpp"
+#include "SqlBaseVisitor.h"
+#include "SqlLexer.h"
+#include "SqlParser.h"
+
+using namespace antlr4;
+using namespace sql;
+
+namespace clp_s::search::sql {
+class ErrorListener : public BaseErrorListener {
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/sql/sql.cpp:17]: yeah it seems like we already have this class in `clp_s::search::kql`

File: components/core/src/clp_s/search/sql/sql.cpp (hunk 6 above)
Comment (gibber9809) [reply components/core/src/clp_s/search/sql/sql.cpp:43]: I'll move the ErrorListener into its own header, and put the visitor classes for sql and kql into the anonymous namespace.

I was worried that BaseErrorListener was part of the generated code for each grammar but it's part of the generic antlr runtime so there shouldn't be any issues.

File: components/core/src/clp_s/search/sql/sql.cpp [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+#include "SqlBaseVisitor.h"
+#include "SqlLexer.h"
+#include "SqlParser.h"
+
+using namespace antlr4;
+using namespace sql;
+using clp_s::search::antlr_common::ErrorListener;
+
+namespace clp_s::search::sql {
+class ErrorListener : public BaseErrorListener {
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/sql/sql.cpp:19]: Should be removed right?

File: components/core/src/clp_s/search/kql/kql.cpp
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/kql/kql.cpp:1]: I will probably leave the change of kql to the next PR. Maybe ask Kirk's advice @kirkrodrigues

File: components/core/src/clp_s/search/antlr_common/ErrorListener.hpp [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+            size_t line,
+            size_t char_position_in_line,
+            std::string const& msg,
+            std::exception_ptr e
+    ) override {
+        m_error = true;
+        m_error_message = msg;
+    }
+
+    bool error() const { return m_error; }
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/antlr_common/ErrorListener.hpp:20]:
```suggestion
    [[nodiscard]] bool error() const { return m_error; }
```

File: components/core/src/clp_s/search/antlr_common/ErrorListener.hpp [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+    std::string const& message() const { return m_error_message; }
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/antlr_common/ErrorListener.hpp:22]:
```suggestion
    [[nodiscard]] std::string const& message() const { return m_error_message; }
```

**Review by LinZhihao-723 — CHANGES_REQUESTED** [review]

**Comment by gibber9809** [comment]
@LinZhihao-723 fixed clang-tidy warnings. Re-review?

File: components/core/src/clp_s/search/sql/sql.cpp [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+using namespace antlr4;
+using namespace sql;
+using clp_s::search::antlr_common::ErrorListener;
+
+namespace clp_s::search::sql {
+namespace {
+class ParseTreeVisitor : public SqlBaseVisitor {
+public:
+    auto visitStart(SqlParser::StartContext* ctx) -> std::any override {
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/sql/sql.cpp:23]:
```suggestion
    [[nodiscard]] auto visitStart([[maybe_unused]] SqlParser::StartContext* ctx) -> std::any override {
```

File: components/core/src/clp_s/search/sql/sql.cpp [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        return EmptyExpr::create();
+    }
+};
+}  // namespace
+
+auto parse_sql_expression(std::istream& in) -> std::shared_ptr<Expression> {
+    ErrorListener lexer_error_listener;
+    ErrorListener parser_error_listener;
+
+    ANTLRInputStream input(in);
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/sql/sql.cpp:33]:
```suggestion
    ANTLRInputStream inpu{in};
```
We should use `{}` initialization whenever possible

File: components/core/src/clp_s/search/sql/sql.cpp [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+    lexer.addErrorListener(&lexer_error_listener);
+    CommonTokenStream tokens(&lexer);
+    SqlParser parser(&tokens);
+    parser.removeErrorListeners();
+    parser.addErrorListener(&parser_error_listener);
+    SqlParser::StartContext* tree = parser.start();
+
+    if (lexer_error_listener.error()) {
+        SPDLOG_ERROR("Lexer error: {}", lexer_error_listener.message());
+        return {};
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/sql/sql.cpp:45]:
```suggestion
        return nullptr;
```
Nit: shall we return `nullptr` explicitly to make it more clear this is an error condition (which matches the docstring)

File: components/core/src/clp_s/search/sql/sql.hpp [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+
+#include "../Expression.hpp"
+
+namespace clp_s::search::sql {
+/**
+ * Parses an SQL expression from the given stream to generate a search AST.
+ * @param in Input stream containing an SQL expression followed by EOF
+ * @return a search AST on success, nullptr otherwise
+ */
+auto parse_sql_expression(std::istream& in) -> std::shared_ptr<Expression>;
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/sql/sql.hpp:15]:
```suggestion
[[nodiscard]] auto parse_sql_expression(std::istream& in) -> std::shared_ptr<Expression>;
```

File: components/core/src/clp_s/search/sql/CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+        ../NullLiteral.hpp
+        ../OrExpr.hpp
+        ../StringLiteral.hpp
+        ${ANTLR_SqlParser_CXX_OUTPUTS}
+        sql.cpp
+        sql.hpp
+)
+target_compile_features(sql PRIVATE cxx_std_20)
+target_include_directories(sql PRIVATE ${ANTLR_SqlParser_OUTPUT_DIR})
+target_link_libraries(sql PRIVATE antlr4_static Boost::filesystem)
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/sql/CMakeLists.txt:29]:
```suggestion
target_link_libraries(sql PRIVATE antlr4_static)
```
Seems like we don't need `Boost::filesystem` now

File: components/core/src/clp_s/search/antlr_common/ErrorListener.hpp [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+class ErrorListener : public antlr4::BaseErrorListener {
+public:
+    void syntaxError(
+            antlr4::Recognizer* recognizer,
+            antlr4::Token* offending_symbol,
+            size_t line,
+            size_t char_position_in_line,
+            std::string const& msg,
+            std::exception_ptr e
+    ) override {
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/antlr_common/ErrorListener.hpp:15]:
```suggestion
    auto syntaxError(
            [[maybe_unused]] antlr4::Recognizer* recognizer,
            [[maybe_unused]] antlr4::Token* offending_symbol,
            [[maybe_unused]] size_t line,
            [[maybe_unused]] size_t char_position_in_line,
            std::string const& msg,
            [[maybe_unused]] std::exception_ptr e
    ) -> void override {
```

File: components/core/src/clp_s/search/antlr_common/ErrorListener.hpp [hunk 17]
```diff
@@ -0,0 +1,30 @@
+#ifndef CLP_S_SEARCH_ANTLRCOMMON_ERRORLISTENER_HPP
+#define CLP_S_SEARCH_ANTLRCOMMON_ERRORLISTENER_HPP
+#include <antlr4-runtime.h>
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/antlr_common/ErrorListener.hpp:3]:
```suggestion

#include <cstddef>
#include <exception>
#include <string>
#include <string_view>

#include <antlr4-runtime.h>
```
Note: string_view is required if the following comment is appiled

File: components/core/src/clp_s/search/antlr_common/ErrorListener.hpp [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+            std::string const& msg,
+            std::exception_ptr e
+    ) override {
+        m_error = true;
+        m_error_message = msg;
+    }
+
+    [[nodiscard]] auto error() const -> bool { return m_error; }
+
+    [[nodiscard]] auto message() const -> std::string const& { return m_error_message; }
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/antlr_common/ErrorListener.hpp:22]:
```suggestion
    [[nodiscard]] auto message() const -> std::string_view { return m_error_message; }
```
How about returning a string view instead?

File: components/core/src/clp_s/search/antlr_common/ErrorListener.hpp [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+
+private:
+    bool m_error{false};
+    std::string m_error_message{};
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/antlr_common/ErrorListener.hpp:26]:
```suggestion
    std::string m_error_message;
```

File: components/core/tests/test-sql.cpp [hunk 20]
```diff
@@ -0,0 +1,23 @@
+#include <sstream>
```
Comment (LinZhihao-723) [inline components/core/tests/test-sql.cpp:1]:
```suggestion
#include <memory>
#include <sstream>
```
For `std::dynamic_pointer_cast`

File: components/core/tests/test-sql.cpp [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+#include "../src/clp_s/search/sql/sql.hpp"
+#include "LogSuppressor.hpp"
+
+using clp_s::search::EmptyExpr;
+using clp_s::search::sql::parse_sql_expression;
+using std::stringstream;
+
+TEST_CASE("Test parsing SQL", "[SQL]") {
+    // Suppress logging
+    LogSuppressor suppressor{};
```
Comment (LinZhihao-723) [inline components/core/tests/test-sql.cpp:16]: Not used?

File: components/core/src/clp_s/search/sql/sql.cpp [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+#include "../antlr_common/ErrorListener.hpp"
+#include "../EmptyExpr.hpp"
+#include "../Expression.hpp"
+#include "SqlBaseVisitor.h"
+#include "SqlLexer.h"
+#include "SqlParser.h"
+
+using namespace antlr4;
+using namespace sql;
+using clp_s::search::antlr_common::ErrorListener;
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/sql/sql.cpp:17]:
```suggestion
using antlr4::ANTLRInputStream;
using antlr4::CommonTokenStream;
using clp_s::search::antlr_common::ErrorListener;
using sql::SqlBaseVisitor;
using sql::SqlLexer;
using sql::SqlParser;
```
We shouldn't add using to the entire namespace; instead we should add using antlr4::xxx to refer to specific used symbols

**Review by LinZhihao-723 — CHANGES_REQUESTED** [review]

File: components/core/tests/test-sql.cpp (hunk 21 above)
Comment (gibber9809) [reply components/core/tests/test-sql.cpp:16]: It's use is entirely implemented in the constructor/destructor. It just turns off spdlog logging once created, and re-enables it once destroyed.

Just allows you to run parsing unit tests without emitting tons of log messages for parsing failure.

File: components/core/src/clp_s/search/sql/Sql.g4
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/sql/Sql.g4:1]: Can we add a comment explaining this is a boilerplate? Asked Kirk to take a brief view and his feedback is this file can be confusing to people new to it without a comment explaining it's a WIP config

File: components/core/tests/test-sql.cpp [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
```
Comment (LinZhihao-723) [inline components/core/tests/test-sql.cpp:17]:
```suggestion
    LogSuppressor const suppressor;
```

File: components/core/tests/test-sql.cpp [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+
+    SECTION("Stub accepts empty string") {
+        stringstream empty_string{""};
+        auto filter = std::dynamic_pointer_cast<EmptyExpr>(parse_sql_expression(empty_string));
+        REQUIRE(nullptr != filter);
```
Comment (LinZhihao-723) [inline components/core/tests/test-sql.cpp:22]:
```suggestion
        REQUIRE((nullptr != filter));
```

File: components/core/tests/test-sql.cpp [hunk 25]
```diff
@@ -0,0 +1,24 @@
+#include <memory>
+#include <sstream>
+
+#include <Catch2/single_include/catch2/catch.hpp>
+#include <spdlog/spdlog.h>
```
Comment (LinZhihao-723) [inline components/core/tests/test-sql.cpp:5]:
```suggestion
```
I think this include should be moved to "LogSuppressor.hpp"

File: components/core/tests/test-sql.cpp (hunk 24 above)
Comment (LinZhihao-723) [reply components/core/tests/test-sql.cpp:22]: Sorry missed this in the previous review. Need to add this to silent clang-tidy

File: components/core/src/clp_s/search/sql/sql.cpp [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+    parser.addErrorListener(&parser_error_listener);
+    SqlParser::StartContext* tree = parser.start();
+
+    if (lexer_error_listener.error()) {
+        SPDLOG_ERROR("Lexer error: {}", lexer_error_listener.message());
+        return nullptr;
+    }
+    if (parser_error_listener.error()) {
+        SPDLOG_ERROR("Parser error: {}", parser_error_listener.message());
+        return {};
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/sql/sql.cpp:53]:
```suggestion
        return nullptr;
```
Let's also be explicit here

File: components/core/src/clp_s/search/sql/sql.cpp [hunk 27]
```diff
@@ -0,0 +1,59 @@
...
+    }
+};
+}  // namespace
+
+auto parse_sql_expression(std::istream& in) -> std::shared_ptr<Expression> {
+    ErrorListener lexer_error_listener;
+    ErrorListener parser_error_listener;
+
+    ANTLRInputStream input{in};
+    SqlLexer lexer(&input);
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/sql/sql.cpp:38]:
```suggestion
    SqlLexer lexer{&input};
```

File: components/core/src/clp_s/search/sql/sql.cpp [hunk 28]
```diff
@@ -0,0 +1,59 @@
...
+    lexer.removeErrorListeners();
+    lexer.addErrorListener(&lexer_error_listener);
+    CommonTokenStream tokens(&lexer);
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/sql/sql.cpp:41]:
```suggestion
    CommonTokenStream tokens{&lexer};
```

File: components/core/src/clp_s/search/sql/sql.cpp [hunk 29]
```diff
@@ -0,0 +1,59 @@
...
+    SqlParser parser(&tokens);
+    parser.removeErrorListeners();
+    parser.addErrorListener(&parser_error_listener);
+    SqlParser::StartContext* tree = parser.start();
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/sql/sql.cpp:45]:
```suggestion
    SqlParser::StartContext* tree{parser.start()};
```

**Review by LinZhihao-723 — CHANGES_REQUESTED** [review]

File: components/core/src/clp_s/search/sql/Sql.g4
Comment (kirkrodrigues) [reply components/core/src/clp_s/search/sql/Sql.g4:1]: Fwiw, the Rabbit agrees with me, lol.

File: components/core/src/clp_s/search/kql/CMakeLists.txt [hunk 30]
```diff
@@ -25,4 +26,4 @@ add_library(
 )
 target_compile_features(kql PRIVATE cxx_std_20)
 target_include_directories(kql PRIVATE ${ANTLR_KqlParser_OUTPUT_DIR})
-target_link_libraries(kql PRIVATE antlr4_static Boost::filesystem)
+target_link_libraries(kql PRIVATE antlr4_static spdlog::spdlog)
```
[the remark this answers is not in the document]
Comment (gibber9809) [reply components/core/src/clp_s/search/kql/CMakeLists.txt:29]: Not relevant to PR.

File: components/core/src/clp_s/search/sql/sql.cpp [hunk 31]
```diff
@@ -0,0 +1,63 @@
...
+    }
+    if (parser_error_listener.error()) {
+        SPDLOG_ERROR("Parser error: {}", parser_error_listener.message());
+        return nullptr;
+    }
+
+    ParseTreeVisitor visitor;
+    try {
+        return std::any_cast<std::shared_ptr<Expression>>(visitor.visitStart(tree));
+    } catch (std::exception& e) {
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/search/sql/sql.cpp:59]:
```suggestion
    } catch (std::exception const& e) {
```

**Review by LinZhihao-723 — CHANGES_REQUESTED** [review]
Should be good to merge after this round.
I will approve after the workflow failure is resolved.

**Comment by kirkrodrigues** [comment]
Can we merge main? The latest workflow failure should be resolved by #578.

**Review by LinZhihao-723 — APPROVED** [review]
For the PR title, how about:
```
feat(clp-s): Add boilerplate for SQL parsing. 
```
